circuits/core/src/algorithms/merkle_path/to_root.rs

```diff
@@ -0,0 +1,158 @@
+
+            curr_hash = crh.hash(&[left_input, right_input].concat());
+        }
+
+        curr_hash
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::algorithms::Pedersen;
+
+    use snarkvm_algorithms::{
+        crh::PedersenCompressedCRH as NativePedersenCompressed,
+        merkle_tree::{MaskedMerkleTreeParameters, MerkleTree},
+        traits::MerkleParameters,
+    };
+
+    use snarkvm_circuits_environment::{assert_scope, Circuit, Mode};
+    use snarkvm_curves::{bls12_377::Fr, edwards_bls12::EdwardsProjective};
+    use snarkvm_utilities::{test_rng, ToBits, UniformRand};
+
+    use std::sync::Arc;
+
+    const PEDERSEN_NUM_WINDOWS: usize = 128;
+    const PEDERSEN_WINDOW_SIZE: usize = 4;
+    const TREE_DEPTH: usize = 4;
+    const MESSAGE: &str = "Pedersen merkle path test";
+
+    type NativeH = NativePedersenCompressed<EdwardsProjective, PEDERSEN_NUM_WINDOWS, PEDERSEN_WINDOW_SIZE>;
+    type Parameters = MaskedMerkleTreeParameters<NativeH, TREE_DEPTH>;
+
+    type H = Pedersen<Circuit, PEDERSEN_NUM_WINDOWS, PEDERSEN_WINDOW_SIZE>;
+
+    fn check_to_root(
+        mode: Mode,
+        use_bad_root: bool,
+        num_constants: usize,
+        num_public: usize,
+        num_private: usize,
+        num_constraints: usize,
+    ) {
+        let merkle_tree_parameters = Parameters::setup(MESSAGE);
+        let crh = H::setup(MESSAGE);
+
+        let mut rng = test_rng();
+        let mut leaves = Vec::new();
+        for _ in 0..1 << Parameters::DEPTH {
+            leaves.push(Fr::rand(&mut rng));
+        }
+
+        let merkle_tree = MerkleTree::new(Arc::new(merkle_tree_parameters), &leaves).unwrap();
+        let root = merkle_tree.root();
+
+        for (i, leaf) in leaves.iter().enumerate() {
+            let proof = merkle_tree.generate_proof(i, &leaf).unwrap();
+            assert!(proof.verify(root, &leaf).unwrap());
+
+            let leaf_bits = leaf.to_bits_le();
+            let root = if use_bad_root { Default::default() } else { *root };
+
+            Circuit::scope(format!("{mode} {MESSAGE} {i}"), || {
+                let traversal = proof.position_list().collect::<Vec<_>>();
+                let path = proof.path.clone();
+                let merkle_path = MerklePath::<Circuit, H>::new(mode, (traversal, path));
+
+                let circuit_leaf = leaf_bits
+                    .iter()
+                    .map(|bit| <H as Hash>::Input::new(mode, *bit))
+                    .collect::<Vec<<H as Hash>::Input>>();
+                let candidate_root = merkle_path.to_root(&crh, &circuit_leaf);
+
+                assert_eq!(*leaf.to_bits_le(), circuit_leaf.eject_value());
+                assert_eq!(root, candidate_root.eject_value());
+
+                let case = format!("mode = {mode}");
+                assert_scope!(case, num_constants, num_public, num_private, num_constraints);
+            });
+        }
+    }
+
+    #[test]
+    fn test_good_root_constant() {
```

Comment:
nit: rename these method to `test_to_root_{mode}` and `test_to_root_{mode}_fails`